prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,675 @@
+ * region that another peer already initialized.
+ *
+ * In either case, we don't need the lock, as we are only growing the file
+ * and not modifying any allocations (ep_entries)
+ *
+ * @param[in,out] map			an sm2_mmap with fid, size and base address
+ * @param[in] last_valid_entry		The id of a region we need to be valid.
+ * @return pointer to new map->base (typically ignored)
+ */
+void *sm2_coordinator_extend_for_entry(struct sm2_mmap *map, int last_valid_entry)
+{
+	size_t new_size;
+	assert(last_valid_entry <= SM2_COORDINATOR_MAX_UNIVERSE_SIZE);
+	new_size = (char *)sm2_mmap_ep_region(map, last_valid_entry + 1) - map->base;
+	return sm2_mmap_remap(map, new_size);
+}
+
+// TODO combine this logic with the remap function while keeping remaps()'s short circuit
+// logic
+static void *sm2_mmap_shrink_to_size(struct sm2_mmap *map, size_t shrink_size)
+{
+	struct stat st;
+	int err;
+
+	err = fstat(map->fd, &st);
+	if (OFI_UNLIKELY(err))
+		goto out;
+
+	if (st.st_size > shrink_size) {
+		/* we need to shrink the file. */
+		FI_WARN(
+		    &sm2_prov, FI_LOG_EP_CTRL,
+		    "sm2_mmap_shrink_to_index() shrinking SHM file to be of size %zu\n",
+		    shrink_size);
+		err = ftruncate(map->fd, shrink_size);
+		if (OFI_UNLIKELY(err))
+			goto out;
+
+		/* now un-map and re-map the file */
+		err = munmap(map->base, map->size);
+		if (OFI_UNLIKELY(err))
+			goto out;
+		map->base =
+		    mmap(0, shrink_size, PROT_READ | PROT_WRITE, MAP_SHARED, map->fd, 0);
+		if (OFI_UNLIKELY(map->base == MAP_FAILED)) {
+			perror("Failed to remap when decreasing the map size in "
+			       "sm2_mmap_shrink_to_size");
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed syscall during sm2_mmap_map(). st.st_size=%ld\n "
+				"shrink_size=%ld",
+				st.st_size, shrink_size);
+			map->base = NULL;
+		}
+		map->size = shrink_size;
+	}
+
+out:
+	if (OFI_UNLIKELY(err)) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"Failed syscall during sm2_mmap_shrink_to_index()\n");
+	}
+	return map->base;
+}
+
+/*
+ * If everything in the file is dead, shrink it to fit 1 entry.
+ * This deals with peer pre-allocated entries by making the assumption that
+ * my PID must be alive in the file in order for me to hold any of my peers allocations.
+ */
+static void sm2_coordinator_attempt_shm_file_shrink(struct sm2_mmap *map)
+{
+	struct sm2_coord_file_header *header = (void *)map->base;
+	struct sm2_ep_allocation_entry *entries = sm2_mmap_entries(map);
+	int jentry;
+	bool shrink_file = true;
+
+	for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+		if (entries[jentry].pid != 0 && pid_lives(abs(entries[jentry].pid))) {
+			shrink_file = false;
+		}
+	}
+
+	if (shrink_file) {
+		/* Reset Map */
+		for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+			entries[jentry].pid = 0;
+			entries[jentry].ep_name[0] = '\0';
+			entries[jentry].startup_ready = 0;
+		}
+
+		/* Shrink File */
+		sm2_mmap_shrink_to_size(map, header->ep_regions_offset);
+	}
+}
```

Comment:
shoot, looks like clang-format doesn't have an easy way to add lines at the end of files.  At least I couldn't find one when I looked.

Reply:
I guess we will have to add manual new lines for now... or maybe we shouldn't care b/c editors should take care of this.

vscode settings.json:

```
{
    "files.insertFinalNewline": true,
    "files.trimFinalNewlines": true
}
```

Note: You will probably get feedback to separate these formatting changes into their own commits.